p2p/rlpx/rlpx.go

```diff
@@ -0,0 +1,408 @@
+		)
+		c.fd.SetDeadline(time.Now().Add(c.cfg.handshakeTimeout()))
+		if c.isServer {
+			vsn, rid, ingress, egress, err = c.recipientHandshake()
+		} else {
+			vsn, ingress, egress, err = c.initiatorHandshake()
+		}
+		if err != nil {
+			return
+		}
+
+		c.mu.Lock()
+		c.vsn = vsn
+		if rid != nil {
+			c.remoteID = rid
+		}
+		c.mu.Unlock()
+		c.rw = newFrameRW(c.fd, ingress, egress)
+		go readLoop(c)
+	})
+	if err == nil && c.rw == nil {
+		return errors.New("handshake failed")
+	}
+	return err
+}
+
+// LocalAddr returns the local network address of the underlying net.Conn.
+func (c *Conn) LocalAddr() net.Addr {
+	return c.fd.LocalAddr()
+}
+
+// RemoteAddr returns the remote network address of the underlying net.Conn.
+func (c *Conn) RemoteAddr() net.Addr {
+	return c.fd.RemoteAddr()
+}
+
+// RemoteID returns the public key of the remote end.
+// If the remote identity is not yet known, it returns nil.
+func (c *Conn) RemoteID() *ecdsa.PublicKey {
+	c.mu.Lock()
+	id := c.remoteID
+	c.mu.Unlock()
+	return id
+}
+
+// Version returns the negotiated RLPx version of the connection.
+// The return value is zero before the handshake has executed and
+// can be 4 or 5 afterwards.
+func (c *Conn) Version() uint {
+	c.mu.Lock()
+	vsn := c.vsn
+	c.mu.Unlock()
+	return vsn
+}
+
+// Close closes the connection.
+func (c *Conn) Close() error {
+	// TODO: shut down reader/wr
+	return c.fd.Close()
+}
+
+// Protocol returns a handle for the given protocol id.
+// It can be called at most once for any given id,
+// subsequent call with the same id will panic.
+func (c *Conn) Protocol(id uint16) *Protocol {
+	p := c.getProtocol(id)
+	close(p.claimSignal) // panics when claimed twice
+	return p
+}
+
+// waits until the given protocol is claimed by a call to Protocol.
+func (c *Conn) waitForProtocol(id uint16) *Protocol {
+	p := c.getProtocol(id)
+	timeout := time.NewTimer(5 * time.Second)
+	defer timeout.Stop()
+	select {
+	case <-timeout.C:
+		return nil
+	case <-p.claimSignal:
+		return p
+	}
+}
+
+func (c *Conn) getProtocol(id uint16) *Protocol {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.proto[id] == nil {
+		c.proto[id] = newProtocol(c, id)
+	}
+	return c.proto[id]
+}
+
+// Protocol is a handle for the given protocol.
+type Protocol struct {
+	c           *Conn
+	claimed     bool
+	id          uint16
+	claimSignal chan struct{}
+
+	// for readLoop
+	xfers       map[uint16]*packetReader
+	readBufSema *bufSema
+
+	// for ReadPacket
+	readCond   *sync.Cond // unblocks ReadPacket
+	newPackets []*packetReader
+	readErr    error
+
+	// for writing
+	contextidSeq uint16
+}
+
+func newProtocol(c *Conn, id uint16) *Protocol {
+	return &Protocol{
+		c:           c,
+		id:          id,
+		claimSignal: make(chan struct{}),
+		xfers:       make(map[uint16]*packetReader),
+		readBufSema: newBufSema(c.cfg.readBufferSize()),
+		readCond:    sync.NewCond(new(sync.Mutex)),
+	}
+}
+
+func (p *Protocol) feedPacket(pr *packetReader) {
+	p.readCond.L.Lock()
```

Comment:
`defer` allocates. If it can be avoided at low cost (like here, where the code does not branch), I'd opt for doing it without the defer.